### PR TITLE
FEATURE: SD-830 implement tooltip format

### PR DESCRIPTION
**Pull request by naibinh**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2826
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2577

# Related Jira tasks
- SD-830: https://jira.intgdc.com/browse/SD-830

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6213/ : FAILURE in 4m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1894/ : SUCCESS in 4s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6224/ : SUCCESS in 21m 13s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1905/ : SUCCESS in 9s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3222/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1906/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2606/ : UNSTABLE in 25m 33s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3225/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1913/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2610/ : SUCCESS in 15m 15s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6229/ : SUCCESS in 5m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1919/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6233/ : SUCCESS in 8m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1924/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6240/ : SUCCESS in 14m 29s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1930/ : SUCCESS in 5s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3235/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1931/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2617/ : SUCCESS in 17m 43s

**Comment by naibinh** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1400.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1933/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2618/ : SUCCESS in 32m 07s

File: src/components/core/geoChart/geoChartTooltip.ts [hunk 1]
```diff
@@ -1,17 +1,35 @@
...
+
+function formatMeasure(item: IGeoTooltipItem, separators?: ISeparators): IGeoTooltipItem {
+    const { title, value, format } = item;
+    return {
+        title,
+        value: isNumber(value) ? numberFormat(value, format, null, separators) : NULL_TOOLTIP_VALUE,
+    };
+}
+
+function formatAttribute(item: IGeoTooltipItem) {
```
Comment (ptquang86) [inline src/components/core/geoChart/geoChartTooltip.ts]: return type is missing :)

File: src/components/core/geoChart/geoChartTooltip.ts [hunk 2]
```diff
@@ -45,10 +74,12 @@ function getTooltipItemHtml(item: IGeoTooltipItem): string {
     }
 
     const { title, value } = item;
+    const escapedValue = isNumber(value) ? value : escape(value);
```
Comment (ptquang86) [inline src/components/core/geoChart/geoChartTooltip.ts]: Please use `isFinite` instead

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6253/ : SUCCESS in 10m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1943/ : SUCCESS in 4s

File: src/components/core/geoChart/geoChartTooltip.ts (hunk 2 above)
Comment (naibinh) [reply src/components/core/geoChart/geoChartTooltip.ts]: Fixed

File: src/components/core/geoChart/geoChartTooltip.ts (hunk 1 above)
Comment (naibinh) [reply src/components/core/geoChart/geoChartTooltip.ts]: Fixed

File: src/components/core/geoChart/geoChartColor.ts [hunk 3]
```diff
@@ -74,25 +72,21 @@ export function getColorPaletteMapping(segmentItems: string[] = []): IObjectMapp
...
         return [];
     }
-    const segmentItemNames: string[] = segmentItems.map(
-        (item: Execution.IResultHeaderItem): string => {
-            const name: string = getHeaderItemName(item);
-            return name || EMPTY_SEGMENT_ITEM;
+    const segmentNames: string[] = segmentValues.map(
+        (value: string): string => {
+            return value || EMPTY_SEGMENT_ITEM;
         },
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/core/geoChart/geoChartColor.ts]: Fixed

File: src/components/core/GeoChart.tsx [hunk 4]
```diff
@@ -147,48 +149,40 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, IGeoCh
         this.setState({ enabledLegendItems });
     };
 
-    private getSegmentSeries(data: Execution.IResultHeaderItem[]): IPushpinCategoryLegendItem[] {
-        const segmentItemsUniq: Execution.IResultHeaderItem[] = unionBy(data, getHeaderItemName);
+    private getCategoryLegendItems(data: string[]): IPushpinCategoryLegendItem[] {
+        const segmentItemsUniq: string[] = union(data); // alphabet sort
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/core/GeoChart.tsx]: Fixed by using `uniq`

File: src/helpers/tests/geoChart/data.spec.ts [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+                    772,
+                    709,
+                    3060,
+                    838,
+                    2844,
+                    1719,
+                    4726,
+                    179,
+                    943,
+                    1005,
```
Comment (ptquang86) [inline src/helpers/tests/geoChart/data.spec.ts]: could you please limit the test data to 10 items only?

File: src/helpers/tests/utils.spec.ts [hunk 6]
```diff
@@ -43,3 +43,31 @@ describe("unwrap", () => {
         expect(unwrap({ key: "value" })).toEqual("value");
     });
 });
+
```
Comment (ptquang86) [inline src/helpers/tests/utils.spec.ts]: Do we need a test for negative number?

File: src/helpers/geoChart/data.ts [hunk 7]
```diff
@@ -1,43 +1,81 @@
...
-    const item = VisualizationObject.isMeasureDefinition(definition) && definition.measureDefinition.item;
-    const uri = isDisplayFormUri(item) && item.uri;
-    return { uri, localIdentifier };
+    if (tooltipTextIndex !== undefined) {
+        geoData[TOOLTIP_TEXT].data = getAttributeData(attributeHeaderItems, tooltipTextIndex);
+    }
+
+    if (sizeIndex !== undefined) {
+        geoData[SIZE].data = getMeasureData(executionResult, sizeIndex);
+        geoData[SIZE].format = getFormatFromExecutionResponse(executionResponse, sizeIndex);
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/helpers/geoChart/data.ts]: `geoData[SIZE]` is holding `name` and `index` so that we should not do it :)

File: src/helpers/geoChart/data.ts [hunk 8]
```diff
@@ -1,43 +1,81 @@
...
+    }
+
+    if (colorIndex !== undefined) {
+        geoData[COLOR].data = getMeasureData(executionResult, colorIndex);
+        geoData[COLOR].format = getFormatFromExecutionResponse(executionResponse, colorIndex);
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/helpers/geoChart/data.ts]: `geoData[COLOR]` is holding `name` and `index` so that we should not do it :)

File: src/helpers/geoChart/data.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ptquang86) [reply src/helpers/geoChart/data.ts]: my bad :))

File: src/helpers/tests/geoChart/data.spec.ts (hunk 5 above)
Comment (naibinh) [reply src/helpers/tests/geoChart/data.spec.ts]: I would cause to update all related storybook images, we should consider to implement it in other separate ticket.

File: src/helpers/tests/utils.spec.ts (hunk 6 above)
Comment (naibinh) [reply src/helpers/tests/utils.spec.ts]: Added

File: src/components/core/GeoChart.tsx [hunk 9]
```diff
@@ -147,48 +149,40 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, IGeoCh
...
-        } = this.props;
+    private setGeoChartInnerOptions(geoData: IGeoData) {
         const { segment } = geoData;
-        const attributeHeaderItems = getGeoAttributeHeaderItems(executionResult, geoData);
-        let segmentData: IPushpinCategoryLegendItem[] = [];
+        let categoryItems: IPushpinCategoryLegendItem[] = [];
         if (segment) {
-            const segmentAttributeHeaderItems = attributeHeaderItems[segment.index];
-            segmentData = this.getSegmentSeries(segmentAttributeHeaderItems);
+            categoryItems = this.getCategoryLegendItems(segment.data as string[]);
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/core/GeoChart.tsx]: Fixed

File: src/helpers/tests/geoChart/data.spec.ts (hunk 5 above)
Comment (ptquang86) [reply src/helpers/tests/geoChart/data.spec.ts]: we can do it later

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6255/ : SUCCESS in 9m 00s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1946/ : SUCCESS in 8s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6267/ : FAILURE in 3m 47s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1958/ : SUCCESS in 4s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3249/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1959/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2627/ : UNSTABLE in 18m 04s

File: src/components/core/geoChart/legends/PushpinSizeLegend.tsx [hunk 10]
```diff
@@ -11,14 +12,16 @@ export interface IPushpinSizeLegendProps {
 }
 
 export default function PushpinSizeLegend(props: IPushpinSizeLegendProps): JSX.Element {
-    const { sizes, format, numericSymbols = [], measureName } = props;
-    const minValue: number = Math.min(...sizes);
-    const maxValue: number = Math.max(...sizes);
+    const { sizes = [], format, numericSymbols = [], measureName } = props;
+    const sizeData = sizes.filter(isFinite);
+    const { min: minValue, max: maxValue } = getMinMax(sizeData);
+
```
Comment (dlcd2203) [inline src/components/core/geoChart/legends/PushpinSizeLegend.tsx]: should we just :
min = Math.min (...sizeData),
....
because I see in getMinMax has duplicated to check isFinite.

**Comment by naibinh** [comment]
extended test - storybook

File: src/components/core/geoChart/legends/PushpinSizeLegend.tsx (hunk 10 above)
Comment (naibinh) [reply src/components/core/geoChart/legends/PushpinSizeLegend.tsx]: Fixed

**Comment by dlcd2203** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6270/ : SUCCESS in 7m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1963/ : SUCCESS in 8s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3251/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/1964/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2629/ : SUCCESS in 22m 40s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3252/ : SUCCESS in 16m 48s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/6272/ : SUCCESS in 16m 05s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1522/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2142/ : SUCCESS in 21m 24s
